crates/test_utils/src/shared_values/mod.rs

```diff
@@ -0,0 +1,1011 @@
+
+            let maybe_previous = data_guard.insert(key.clone(), value);
+
+            for (pattern, notifier) in self.notification.lock().await.iter() {
+                if key.matches(pattern).count() > 0 {
+                    tracing::debug!("{key} matched by {pattern}");
+                    notifier.notify_waiters();
+                } else {
+                    tracing::debug!("{key} not matched by {pattern}");
+                }
+            }
+
+            Ok(maybe_previous)
+        }
+    }
+
+    #[async_trait]
+    impl SharedValues for LocalV1 {
+        async fn put_t(&mut self, key: String, value: String) -> Fallible<Option<String>> {
+            self.put(key, value).await
+        }
+
+        async fn get_pattern_t(
+            &mut self,
+            pattern: String,
+            min_data: usize,
+            maybe_wait_timeout: Option<Duration>,
+        ) -> Fallible<Data<String>> {
+            tokio::select! {
+                data = self.get_pattern(pattern.as_str(), |(_previous_data, data)| { data.len() >= min_data }) => Ok(data?),
+                _ = {
+                    let duration = if let Some(wait_timeout) =  maybe_wait_timeout {
+                        wait_timeout
+                    } else {
+                        std::time::Duration::MAX
+                    };
+
+                    tokio::time::sleep(duration)
+                }  => {
+                    anyhow::bail!("timeout")
+                }
+            }
+        }
+        async fn num_waiters_t(&self) -> Fallible<usize> {
+            Ok(self.num_waiters())
+        }
+    }
+}
+
+pub(crate) mod remote_v1 {
+    use anyhow::{bail, Context, Result as Fallible};
+    use async_trait::async_trait;
+    use futures::StreamExt;
+
+    use serde::{Deserialize, Serialize};
+    use std::sync::Arc;
+    use std::time::Duration;
+    use tokio::task::JoinHandle;
+
+    use holochain_websocket::{WebsocketConfig, WebsocketListener};
+
+    // TODO: deglob this
+    use holochain_serialized_bytes::prelude::*;
+
+    use super::local_v1::LocalV1;
+    use super::{Data, SharedValues};
+
+    pub(crate) const SHARED_VALUES_REMOTEV1_URL_ENV: &str = "TEST_SHARED_VALUES_REMOTEV1_URL";
+    pub(crate) const SHARED_VALUES_REMOTEV1_URL_DEFAULT: &str = "ws://127.0.0.1:0";
+
+    // The value given to this env var is used to construct `RemoteV1Role`
+    pub(crate) const SHARED_VALUES_REMOTEV1_ROLE_ENV: &str = "TEST_SHARED_VALUES_REMOTEV1_ROLE";
+
+    pub(crate) const TEST_AGENT_READINESS_REQUIRED_AGENTS_ENV: &str =
+        "TEST_AGENT_READINESS_REQUIRED_AGENTS";
+    pub(crate) const TEST_AGENT_READINESS_REQUIRED_AGENTS_DEFAULT: usize = 3;
+    pub(crate) const TEST_AGENT_READINESS_TIMEOUT_SECS_ENV: &str =
+        "TEST_AGENT_READINESS_TIMEOUT_SECS";
+    pub(crate) const TEST_AGENT_READINESS_TIMEOUT_SECS_DEFAULT: u64 = 360;
+
+    #[derive(Debug, Default, strum_macros::EnumString)]
+    #[strum(serialize_all = "lowercase")]
```

Comment:
Ahh this is neat, I could do with adding `strum` to my toolkit then because I usually do this by hand :D